frappe/core/doctype/communication/communication.py

```diff
@@ -377,6 +377,7 @@ def deduplicate_timeline_links(self):
 
 	def add_link(self, link_doctype, link_name, autosave=False):
 		self.append("timeline_links", {"link_doctype": link_doctype, "link_name": link_name})
+		self.deduplicate_timeline_links()
```

Comment:
Can cause unnecessary recursive call since `deduplicate_timeline_links` is calling `add_link`?

Simply check if link already exists before appending? (we can even get rid of deduplicate_timeline_links after this)